File: 0013. Roman to Integer/solution.py
class Solution:
    # I             1
    # V             5
    # X             10
    # L             50
    # C             100
    # D             500
    # M             1000
    NUMERAL_MAP: dict[str, int] = {
        "I": 1,
        "V": 5,
        "X": 10,
        "L": 50,
        "C": 100,
        "D": 500,
        "M": 1000
    }

    def romanToInt(self, s: str) -> int:
        if not s:
            return 0

        if len(s) == 1:
            return self.NUMERAL_MAP[s[0]]

        if self.NUMERAL_MAP[s[0]] < self.NUMERAL_MAP[s[1]]:
            return (self.NUMERAL_MAP[s[1]] - self.NUMERAL_MAP[s[0]]) + self.romanToInt(s[2:])

        return self.NUMERAL_MAP[s[0]] + self.romanToInt(s[1:])

File: 0013. Roman to Integer/solution.c
#include <string.h>


int numeral_to_int(char numeral)
{
    switch (numeral)
    {
        case 'I':   return 1;
        case 'V':   return 5;
        case 'X':   return 10;
        case 'L':   return 50;
        case 'C':   return 100;
        case 'D':   return 500;
        case 'M':   return 1000;
    }

    return 0;
}

int romanToInt(char* s)
{
    if (strlen(s) == 0)
        return 0;

    if (strlen(s) == 1)
        return numeral_to_int(s[0]);

    int num0 = numeral_to_int(s[0]);
    int num1 = numeral_to_int(s[1]);
    if (num0 < num1)
        return (num1 - num0) + romanToInt(s + 2);

    return numeral_to_int(s[0]) + romanToInt(++s);
}
